src/lib.rs

```diff
@@ -0,0 +1,512 @@
+#[derive(thiserror::Error, PartialEq)]
+pub enum Error {
+    #[error("configuration error: {0}")]
+    ConfigError(String),
+    #[error("API error: {0}")]
+    ApiError(String),
+    #[error("callback error: {0}")]
+    CallbackError(String),
+    #[error("feature not implemented: {0}")]
+    NotImplementedError(String),
+}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::NotImplementedError(e)
+            | Error::ConfigError(e)
+            | Error::ApiError(e)
+            | Error::CallbackError(e) => {
+                write!(f, "{}", e)
+            }
+        }
+    }
+}
+
+/// EvidenceBuilder is the function signature of the application callback.
+/// The application is passed the nonce and the list of supported evidence media
+/// types and shall returns the computed evidence together with the selected
+/// media type.
+type EvidenceBuilder =
+    fn(nonce: &Vec<u8>, accepted: &Vec<String>) -> Result<(Vec<u8>, String), Error>;
+
+pub struct ChallengeResponseConfig {
+    nonce: Option<Vec<u8>>,
+    nonce_sz: usize,
+    base_url: Option<url::Url>,
+    http_client: Option<reqwest::blocking::Client>,
+    delete_session: bool,
+    evidence_builder: Option<EvidenceBuilder>,
+}
+
+const CRS_MEDIA_TYPE: &'static str = "application/vnd.veraison.challenge-response-session+json";
+
+impl ChallengeResponseConfig {
+    pub fn new() -> Self {
+        Self {
+            nonce: None,
+            nonce_sz: 0,
+            base_url: None,
+            delete_session: true,
+            http_client: None,
+            evidence_builder: None,
+        }
+    }
+
+    pub fn set_base_url(&mut self, v: String) -> Result<(), Error> {
+        let url = url::Url::parse(&v).map_err(|e| Error::ConfigError(e.to_string()))?;
+
+        self.base_url = Some(url);
+
+        Ok(())
+    }
+
+    pub fn set_nonce(&mut self, v: Vec<u8>) -> Result<(), Error> {
+        if v.len() == 0 {
+            return Err(Error::ConfigError("zero length nonce supplied".to_string()));
+        }
+
+        self.nonce = Some(v);
+        self.nonce_sz = 0;
+
+        Ok(())
+    }
+
+    pub fn set_nonce_sz(&mut self, v: usize) -> Result<(), Error> {
+        if v == 0 {
+            return Err(Error::ConfigError("zero bytes nonce requested".to_string()));
+        }
+
+        self.nonce_sz = v;
+        self.nonce = None;
+
+        Ok(())
+    }
+
+    pub fn set_delete_session(&mut self, v: bool) {
+        self.delete_session = v;
+    }
+
+    pub fn set_http_client(&mut self, v: reqwest::blocking::Client) {
+        self.http_client = Some(v);
+    }
+
+    pub fn set_evidence_builder(&mut self, v: EvidenceBuilder) {
+        self.evidence_builder = Some(v);
+    }
+
+    // on success return the attestation result
+    pub fn run(&self) -> Result<String, Error> {
```

Comment:
How is that attestation result supposed to be encoded in this `String`? Would it be worth creating some new type for this that can be used more easily, or is it way too specific and the user of the library will have to have their own tooling for that?

Reply:
my plan was to have a dedicated library to interpret the attestation result.  So maybe a better format for this is byte string.